service/k8sclient/k8sclient.go

```diff
@@ -2017,3 +1996,248 @@ func vmAgentSpec(params *PMM, secretName string) monitoring.VMAgent {
+	req := &psmdb.PerconaServerMongoDB112{
+		APIVersion: c.getAPIVersionForPSMDBOperator(extra.operators.PsmdbOperatorVersion),
+		Kind:       psmdb.PerconaServerMongoDBKind,
+		TypeMeta: common.TypeMeta{
+			APIVersion: c.getAPIVersionForPSMDBOperator(extra.operators.PsmdbOperatorVersion),
+			Kind:       psmdb.PerconaServerMongoDBKind,
+		},
+		ObjectMeta: common.ObjectMeta{
+			Name:       params.Name,
+			Finalizers: []string{"delete-psmdb-pvc"},
+		},
+		Spec: &psmdb.PSMDB112Spec{
+			UpdateStrategy: updateStrategyRollingUpdate,
+			CRVersion:      extra.operators.PsmdbOperatorVersion,
+			Image:          extra.psmdbImage,
+			Secrets: &psmdb.SecretsSpec{
+				Users: extra.secretName,
+			},
+			Sharding: &psmdb.ShardingSpec112{
+				Enabled: true,
+				ConfigsvrReplSet: &psmdb.ReplsetSpec{
+					Size:       3,
+					VolumeSpec: c.volumeSpec(params.Replicaset.DiskSize),
+					Arbiter: psmdb.Arbiter{
+						Enabled: false,
+						Size:    1,
+						MultiAZ: psmdb.MultiAZ{
+							Affinity: extra.affinity,
+						},
+					},
+					MultiAZ: psmdb.MultiAZ{
+						Affinity: extra.affinity,
+					},
+					Expose: extra.expose,
+				},
+				Mongos: &psmdb.ReplsetMongosSpec112{
+					Size: params.Size,
+					MultiAZ: psmdb.MultiAZ{
+						Affinity: extra.affinity,
+					},
+					Expose: psmdb.ExposeSpec{
+						ExposeType: common.ServiceTypeLoadBalancer,
+					},
+				},
+			},
+			Replsets: []*psmdb.ReplsetSpec112{
+				{
+					Name:      "rs0",
+					Size:      params.Size,
+					Resources: c.setComputeResources(params.Replicaset.ComputeResources),
+					Arbiter: psmdb.Arbiter{
+						Enabled: false,
+						Size:    1,
+						MultiAZ: psmdb.MultiAZ{
+							Affinity: extra.affinity,
+						},
+					},
+					VolumeSpec: c.volumeSpec(params.Replicaset.DiskSize),
+					PodDisruptionBudget: &common.PodDisruptionBudgetSpec{
+						MaxUnavailable: pointer.ToInt(1),
+					},
+					MultiAZ: psmdb.MultiAZ{
+						Affinity: extra.affinity,
+					},
+					Configuration: "      operationProfiling:\n" +
+						"        mode: " + string(psmdb.OperationProfilingModeSlowOp) + "\n",
+				},
+			},
+			PMM: &psmdb.PmmSpec{
+				Enabled: false,
+			},
+			Backup: &psmdb.BackupSpec{
+				Enabled:            true,
+				Image:              extra.backupImage,
+				ServiceAccountName: "percona-server-mongodb-operator",
+			},
+		},
+	}
+
+	if params.Replicaset != nil {
+		req.Spec.Replsets[0].Resources = c.setComputeResources(params.Replicaset.ComputeResources)
+		req.Spec.Sharding.Mongos.Resources = c.setComputeResources(params.Replicaset.ComputeResources)
+	}
+	if params.PMM != nil {
+		req.Spec.PMM = &psmdb.PmmSpec{
+			Enabled:    true,
+			ServerHost: params.PMM.PublicAddress,
+			Image:      pmmClientImage,
+			Resources: &common.PodResources{
+				Requests: &common.ResourcesList{
+					Memory: "300M",
+					CPU:    "500m",
+				},
+			},
+		}
```

Comment:
I didn't remove it because the methods `c.makeReq` and `c.makeReq112Plus` return an `interface{}`.
To extract this code outside the make method I would need to add a type switch and it would be duplicated anyway.

Reply:
we can create a method which returns PmmSpec